### PR TITLE
rthreads work from desmume

### PR DESCRIPTION
**Pull request by zeromus**
- port slock and scond fixes from desmume
- make sthread_isself return false for null test thread
full disclosure: this work was done months ago. I can't be 100% sure I've merged it correctly with recent rthreads.c changes

### CODE REVIEW

**Comment by twinaphex** [comment]
Inviting @Alcaro and @bparker06 to review this. Maybe @bparker06 can test it as well.

**Comment by zeromus** [comment]
> * pthread_cond_signal doesn't require a specific thread to be woken; drop that requirement and you don't need the potato and queue, making everything a lot simpler and faster (lock stealing isn't wrong, last to enter has hotter cache)

Consider the case where one thread is managing several tasks. We need to ensure there's some mechanism to let other conditions waken instead of the same one over and over. There _is_ a (practical, if unwritten) requirement that the list of waiters progress eventually, so we need some kind of list here--the scheduling policy may permit one waiter to get woken repeatedly, but it is an invalid scheduling policy which never wakes any other waiter.  For the same one-thread reason, even though we don't strictly need a queue, we need to make sure the waiter that just started waiting lets other waiters proceed first (else only the first waiter would proceed) and so he must be wakened after at least one other waiter; a queue with just-wakened entered at the end is the most straightforward way of implementing this.

> * Why is slock_t not a CRITICAL_SECTION, they're faster
> * Mixing SignalObjectAndWait and slock_lock is confusing

Because I use it as an event with SignalObjectAndWait. I think I may not need to do that. It may be safe to do with a CRITICAL_SECTION after all. I'll consider this.

> * pthread_cond_wait is allowed to have spurious wakeups, the scond_new comments suggest the opposite

OK, you got me. This is tricky. Here's how I'm going to rewrite the comment:

`The main problem is that a condition variable can't be used to "pre-wake" a thread (it will get wakened only after it's waited) whereas a win32 event can pre-wake a thread (the event will be set in advance, so a 'waiter' won't even have to wait on it). Keep in mind a condition variable can apparently pre-wake a thread, insofar as spurious wakeups are always possible, but nobody will be expecting this and it does not need to be simulated.`

Also -- if this is so important (and I know it is) then why did scond_wait ditch pthread_cond_wait's return value? That should probably be fixed too.  In desmume we're not using the return value because we have other ways (I dont understand it anymore) to detect spurious wakeups.

> * pthread_cond_signal does not require the lock held, yet the scond_signal comments assume it does; if you want to deviate from POSIX semantics, this must be clearly called out

This is hugely stupid and deeply troubling.  I think this means I need a static critical section to control the variables scond_signal and scond_wait manage between themselves, since we can't trust that the scond_signal caller is using the lock at all. I'll consider this

**Comment by Alcaro** [comment]
> We need to ensure there's some mechanism to let other conditions waken instead of the same one over and over. [...] it is an invalid scheduling policy which never wakes any other waiter

Are you assuming the worst-case scheduler?

If yes, the kernel is allowed to wake only the same thread (or the same two threads) over and over on the potato, yielding a deadlock (or livelock or something); I think you'd need two potato events to avoid that.

If no, any non-hostile scheduler will eventually wake all threads.

> Because I use it as an event with SignalObjectAndWait.

That one isn't documented as atomic, it seems to only save a few context switches if the thread is going to sleep anyways. CRITICAL_SECTION is handled in userspace unless contended, so there's no performance savings there.

> but nobody will be expecting this and it does not need to be simulated.

That means we don't create additional spurious wakeups, but it also means we don't spend any effort on removing them.

I think you're assigning condvars different semantics than everyone else. A condition variable is associated with a boolean predicate telling whether the thread should continue, and this predicate should be checked after every return; spurious wakeups look like real ones. This is called out in the documentation of [SleepConditionVariableCS/SRW](https://msdn.microsoft.com/en-us/library/windows/desktop/ms686301(v=vs.85).aspx), [java.lang.Object.wait()](https://docs.oracle.com/javase/7/docs/api/java/lang/Object.html#wait()), and [pthread_cond](http://pubs.opengroup.org/onlinepubs/9699919799/functions/pthread_cond_wait.html); they all mention spurious wakeups, yet their only documented errors are "your mutex is broken", "bad arguments" and "timeout". ([futex](http://man7.org/linux/man-pages/man2/futex.2.html) does return whether the wakeup is spurious, but it recommends ignoring that and checking the predicate, like the others.)

> why did scond_wait ditch pthread_cond_wait's return value?

Because, as mentioned, it doesn't tell whether it's spurious.

I'm more interested in why scond_signal/broadcast don't return void. Their underlying functions [have no documented errors whatsoever](http://pubs.opengroup.org/onlinepubs/9699919799/functions/pthread_cond_broadcast.html), and the only error they ever had was "that's not a condvar", which is an application bug that we usually delegate to the segfault handler.

> I think this means I need a static critical section to control the variables scond_signal and scond_wait manage between themselves, since we can't trust that the scond_signal caller is using the lock at all.

Why not put that in the condvar? There's no need to synchronize anything between multiple cvs.

**Comment by zeromus** [comment]
> Are you assuming the worst-case scheduler?
> 
> If yes, the kernel is allowed to wake only the same thread (or the same two threads) over and over on the potato, yielding a deadlock (or livelock or something); I think you'd need two potato events to avoid that.
> 
> If no, any non-hostile scheduler will eventually wake all threads.

Gack, I messed up earlier. Consider the case where one _core_ is managing multiple threads. That's what I meant. That is the worst case I'm assuming. In that case we need to get the kernel to switch threads on the one core in a fair way, and the events are needed for that. The kernel is fair (as long as we GET to the kernel) so we only need one potato. But without events we never get to the kernel.

None of this is hypothetical. We were fixing actual hangs in desmume when working on this.

> That one isn't documented as atomic

That's right. It seems safe to do it with a critical section lock and then a signal on the potato.

> That means we don't create additional spurious wakeups, but it also means we don't spend any effort on removing them.

I'm not trying to remove them. But we've already discussed that. You think I'm trying to remove spurious wakeups (hot threads repeatedly hammering the synchronization resources). I'm not -- I'm trying to remove one core hammering the synchronization resources without a preemption to another thread.

As to my trying to assign different semantics to condvars, that's ridiculous. I've been trying extraordinarily hard to achieve the correct semantics.

> > why did scond_wait ditch pthread_cond_wait's return value?

> Because, as mentioned, it doesn't tell whether it's spurious.

Ah, I misread the docs. OK. No need to know whether it's spurious from the return value, no useful information lost



> > I think this means I need a static critical section to control the variables scond_signal and scond_wait manage between themselves, since we can't trust that the scond_signal caller is using the lock at all.

> Why not put that in the condvar? There's no need to synchronize anything between multiple cvs.

OK, I think you're right.

**Comment by Alcaro** [comment]
> we need to get the kernel to switch threads

The kernel does switch threads even without syscalls, all modern systems are preemptive. But they don't switch fast enough to keep a 60fps framerate, if you synchronize every frame; Windows default timer frequency is 16ms (64fps), and I don't think RetroArch reduces that.

But I doubt making a few syscalls would help. I expect the thread calling SetEvent to keep running, since its cache is hotter than anything else the kernel could wake; it will then steal its own potato a few thousand times until the kernel throws the timeout (and then there's a fair chance the interrupted thread holds the mutex and/or hasn't put the potato back, so it gets to run a few more times).

Instead, to force another schedulable thread to run, call Sleep(0).


But you're right on one thing. The current implementation is broken, and my proposed simplifications have the same problem: If a pthread_cond_signal triggers a future pthread_cond_wait on the same thread, then that wakeup is spurious (which is suboptimal, but acceptable). If there were other sleepers during this, that thread should've been woken instead; missing that is a violation of the cv specification.

I tried keeping track of the latest thread to call signal() and enforce an extra wait on that:
```
condvar :=
  int last_waker := -1
  event ev
  int waiters := 0 # these two ints are needed for broadcasts only
  int wakers := 0
  event ev_extra

  private unlock_wait(lock, event):
    lock.unlock()
    event.wait()
    lock.lock()

  wait(lock):
    waiters++
    if last_waker == thread_id():
      unlock_wait(lock, ev_extra)
    # the above was probably the one from our own signal(), let's check again
    # but there can only be one of those, so no need for loops
    if last_waker == thread_id():
      unlock_wait(lock, ev_extra)
    unlock_wait(lock, ev)
    wakers--
    if wakers > 0:
      ev.signal()
    waiters--

  signal():
    last_waker := thread_id()
    wakers++
    ev.wake()
    ev_extra.wake()

  broadcast():
    wakers := waiters-1
    signal()
```
But (1)wait / (2)wait / (3)broadcast / (1)wait steals t2's wakeup. While I still don't like the potato dance, I can't offer any simpler alternative, so your proposal is the best option we have.

Perhaps some variant using a wakeup counter (to allow multiple woken threads to run in any order) and/or two potatoes (to avoid the potentially-unbounded self-bouncing) would be worth it. But this is good enough; keeping the threads waiting for the right potato would get annoying quite fast, and dealing with integer overflow is no fun either.

> As to my trying to assign different semantics to condvars, that's ridiculous. I've been trying extraordinarily hard to achieve the correct semantics.

To achieve correct behavior, you must first know exactly what the correct behavior is. You're not intentionally breaking the cv semantics, but if your understanding of condvars is nonstandard (such as spurious wakeups being detectable), you're likely to accidentally create incorrect cv semantics.

But while your posts and comments are a bit strange, the code seems correct; you have clearly thought more about this than I have.

(On a totally unrelated note, another possible cleanup: Add a dummy head node to the linked list (just remove the asterisk on line 108) and you can get rid of that asterisk-ampersand maze on line 417.)

And I have one final question:

Why are condition variables so popular? They are (as demonstrated) a huge pain to implement, and (with the spurious wakeups and mandatory mutex) fairly complex to use. What advantages do they provide over an event or semaphore?

**Comment by zeromus** [comment]
Yes of course, windows switches eventually. That isn't good enough. Making a few syscalls DOES help. All you need to do is bail out of your thread so windows can wake up another thread on the core. But Sleep(0) isn't enough. Windows needs to know that the current thread is BORED and someone else needs to run. Yeah, you say it's perfectly fine for that thread to get control right back. But that isn't perfectly fine. Mutexes are describing precisely what we want: we want another thread to take over now. Anything other than that is not what we want. So what if it's legal? This is how we tell the OS to HURRY and not do something technically OK suboptimal.

I dont know why condition variables are so popular. I think they're lame. I'll tell you my theory. They're popular because everyone uses them because thats what pthreads offers because that was what was convenient in its original implementation. That Windows should come up with something so completely different is evidence there was multiple ways to implement all this; we're left with a standard one might assume was 'ideal' if they didnt have my theory that it was just how it happened to be.

Using pthreads makes a certain amount of sense for the average person but somewhat less sense for retroarch given its demands to run on more bizarre environments. Has pthreads been safely simulated on all those environments? I'd be concerned. It would take some more detailed information-gathering, but I'd say theres at least a chance that it would be more prudent to BURY the pthreads type primitives under an entirely different wrapper (designed how? for maximum portability, at least.) Guess its time to make a synchronization library from scratch using nothing but OS threads and atomic operations...

I'll apply all the mentioned cleanups and fixes (critical section in place of one mutex) shortly, except foo=bar=0 which seems innocent to me and since you're not sure anyway...

**Comment by Alcaro** [comment]
> But Sleep(0) isn't enough. Windows needs to know that the current thread is BORED and someone else needs to run.

Let me quote the Sleep() docs:

> dwMilliseconds [in]
> A value of zero causes the thread to relinquish the remainder of its time slice to any other thread of equal priority that is ready to run. If there are no other threads of equal priority ready to run, the function returns immediately, and the thread continues execution.

Try it. Perhaps it helps, perhaps it does nothing, but it makes your intentions more clear than "uh let's bounce this event for a while".

Perhaps the Windows kernel does sometimes reschedule threads during syscalls. Perhaps it reschedules only during SignalObjectAndWait. Windows does a lot of strange stuff.

> I'll tell you my theory. They're popular because everyone uses them because thats what pthreads offers because that was what was convenient in its original implementation.

I doubt cv was very convenient even back then. And it's not that it was first, either; the semaphore is ten years older.

The bigger question is why the C11/C++11 guys chose to implement only mutex, condvar, call_once and promise (latter two being C++ only), they're big enough to go against the flow. I googled around a bit for the proposals, and found [N2320](http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2007/n2320.html#Condition) and [N2497](http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2008/n2497.html#thread.condition); they specify that condition variables should be implemented in C++, but not why.

At least for C, this is the wrong choice; C should provide the lowest-level devices and let the user build stuff on top of that, not enforce a specific way to think. Event and semaphore are definitely lower level than cv.

> Has pthreads been safely simulated on all those environments?

Judging by the amount of ifdefs on the headers compared to slock_lock/scond_signal/etc, I'd say yes. Judging by you implementing cv on top of mutex/event, I'd say yes.

> Guess its time to make a synchronization library from scratch using nothing but OS threads and atomic operations...

Sure, why not, it's not like we're sane around here. (Though atomics are strangely absent from our headers...)

Or at least add event to rthreads, so we can replace RETRO_WIN32_USE_PTHREADS with RETRO_NATIVE_CONDVAR and isolate the Windows-specific calls.

If anyone adds event to rthreads (feel free to base it on our semaphore-on-condvar thingy) prior to March 1, I'll contribute a Linux futex backend. (This is out of scope for this PR, of course.)

**Comment by zeromus** [comment]
First I want to be clear that I hadn't ever thought of using Sleep. My instincts are biased against it, both from a "sleep is usually never the right approach" formula and practical experience.

Now, that I've thought about it: in the event one of the threads had a different priority (let's say there were three high priority threads and one low priority thread) then the two high priority threads could sleep back and forth while starving the low priority thread -- because they're not using all the tools at their disposal to make their true requirements known to the OS.  Admittedly this is not the case in desmume.  But I hope from this example you can see how Sleep is an inferior description of the situation.

**Comment by Alcaro** [comment]
Sleep is indeed not perfect. Even without thread priorities, it's possible that two threads bounce the potato forever (it'd require a really dumb scheduler, of course, but it's possible). Nonzero sleep is nothing more than an admission that we couldn't do it properly.

The only solutions that guarantee timely rescheduling would require two potatoes, or something like that, and that's a bit more effort than it's worth.

But I still believe Sleep(0) is better than nothing. Maybe not much better, but at least a little.

**Comment by zeromus** [comment]
So.. I guess the scenario I just gave where sleep is insufficient is also insufficient for one potato. 
Solving this scenario is getting much too close to making our own scheduler.
Therefore it's possible we could proceed testing with Sleep instead of the potatos. 
But, I'd rather not. I want this case closed. What we have is an improvement over what we had before.
How about I add a comment about sleep being an alternative to potatos and link to this thread and if anyone ever finds bugs in the potato we can address it again with simpler code?

**Comment by Alcaro** [comment]
There's no need for any big rewrite, just call Sleep(0) between releasing the mutex and awaiting the potato.

I suspect it's possible to rewrite it to only one event, but I'm not sure that improves anything.

If we want a rewrite, we should aim for pingponging the threads between two events, to ensure they all get their turn no matter what the scheduler tries to do.

And that list should become wakeup counters; one integer for number of wakeups, another for number of waits. Each call copies and increments the wait counter, and leaves once wakeups >= their ID (modulo integer overflows; we can safely assume nothing can sleep through 2^31 wakeups). If two threads should wake up, which of them goes first doesn't matter.

If this PR (or an equivalent) is merged, and events are added to rthreads, prior to February 11, I'll implement these suggestions. I will also check a few other cv implementations and see if they do anything I should steal.

@bparker06, do you have anything to add?

**Comment by zeromus** [comment]
Oh, I get it now. the Sleep is designed to keep the current thread from getting the potato back immediately. It does inform the OS of our true intentions (we DID mean to send the potato to someone else. we expect to be bored presently waiting for someone else to finish).

I'm not interested in a rewrite. This code solved our bugs. I've gone way beyond my enthusiasm budget on this. The next-level enthusiasm is writing an entire replacement API or handmade scheduler. I dont have the enthusiasm now for that. But.. like I said.. it's a thing I think is within this library's purview.

**Comment by twinaphex** [comment]
Alright, what is the status on this? Can we merge this, does it still require work, etc.?

**Comment by Alcaro** [comment]
Not ready for merge. I'll assume bparker has nothing to add, but we still found a bunch of stuff to fix.

**Comment by zeromus** [comment]
I'm stymied. Changing the first SignalObjectAndWait(A,B) to a SetEvent(A) and WaitForSingleObject(B) results in deadlocks. I can't figure out why that would happen. As you expected, there should need to be some kind of yield (you were suggesting a Sleep) between those two calls, but everything I tried kept deadlocking. Is there some flaw in the algorithm that requires SignalObjectAndWait to be atomic (even though it's not guaranteed)? Even a Sleep(10) deadlocks (while running slow).

**Comment by bparker06** [comment]
When is USE_WIN32_THREADS even defined? I cannot find any reference to a definition of it in the code.

**Comment by Alcaro** [comment]
Unless you accidentally changed the algorithm since I reviewed it, it should be safe. If it did depend on that, it should've deadlocked long ago.

Is there any chance you can replace the desmume condvars with some less uncooperative object? Or at least grab stack traces/etc of the deadlock? :/

> When is USE_WIN32_THREADS even defined?

Line 35 of that file, he didn't change it.

But it depends on RETRO_WIN32_USE_PTHREADS, which is completely absent. Not sure what the plan is.

**Comment by zeromus** [comment]
I found the problem. `Changing the first SignalObjectAndWait(A,B) to a SetEvent(A) and WaitForSingleObject(B)` is a bug! Instead, I need to `Change SignalObjectAndWait(A,B) to a ReleaseMutex(A) and WaitForSingleObject(B)`.

With this commit I believe I applied all the feedback

Wait a while longer, I'm gonna leave it in desmume to smoke out any other surprises

**Comment by Alcaro** [comment]
ohh... yeah, of course it dies if wrong function is used. (pthreads would've segfaulted. It's a bit rude, but at least we'd get a clear stack trace.)

I found the following complaints; nothing critical, but they would be nice.
- Mismatched paren on line 448
- Lines 353-360: Mention that self-wakeups on the event are spurious and therefore fine, but they steal the real wakeup, which is less fine
- scond_signal is still a race condition if lock isn't held
- I'm not completely sure scond_wait_timeout works at all with the current setup; spurious wakeups are fine, but it may eat the event signals in unexpected ways (not sure, didn't check)
- (The simplifications/rewrites I mentioned above are more effort than they're worth.)

**Comment by zeromus** [comment]
I addressed all the new issues except for scond_wait_timeout. 

I need to think about this more.

I believe I should implement it by making the basic new algorithm accept a timestamp (possibly infinite) and manage the timeout internally on its own. This won't be trivial... and I don't have any means of testing it now...

I'll have to make one or more of the relevant waits have a timeout on it, at which point the total duration for the operation will be evaluated, and the operation can bail itself out (I'll have to think of the mechanism for doing that-- removing from the middle of the queue... grumble...) at that time.

This PR may sit here another month if I have to do all this.

**Comment by Alcaro** [comment]
> \+   /* Moreover, we won't be doing this, because it counts as a spurious wakeup -- someone else with a genuine claim must get wakened, in any case.
> \+   /* Therefore we choose to wake only one of the correct waiting threads. */

Missing \*/ (not sure why you're using that style for this comment rather than same as copyright header, but sure, why not)

> removing from the middle of the queue... grumble...

Broadcast the cv and call it spurious?

Or the slightly simpler solution:

> \+ #ifndef _WIN32 /* we couldn't get this working */
> bool scond_wait_timeout(scond_t *cond, slock_t *lock, int64_t timeout_us);
> \+ #endif

I'm fine with merging this, but you said you want to wait some more, so let's do that.

**Comment by twinaphex** [comment]
Let's not wait an entire month though, that is really pushing things. I'd like for libretro-common in desmume and retroarch/upstream to be on the same page again as soon as possible so I can return to that core.

So let's try to find a compromise here that nevertheless is stable.

**Comment by bparker06** [comment]
Just FYI SignalObjectAndWait is not available on Windows 95 (which is already used in the existing code), which I need to find a workaround for already. So if this is merged then a similar workaround will need to be used. Any ideas on an alternative?

**Comment by Alcaro** [comment]
SignalObjectAndWait(a,b); == SignalObject(a); WaitForSingleObject(b); where SignalObject is one of ReleaseMutex, SetEvent, or a few others. I believe the object type (and thereby the replacement function) is known everywhere it's used.

Don't worry about it, this PR removes them all.

**Comment by bparker06** [comment]
@Alcaro The PR doesn't remove the usage of SignalObjectAndWait though, it actually adds another one.

**Comment by Alcaro** [comment]
The first commit adds another one. Later commits remove them all.

**Comment by bparker06** [comment]
Ah I only saw the first one. Thanks

**Comment by twinaphex** [comment]
What is the status on this now? I'd like to fast-track this but I also want to make sure it won't break anything.

**Comment by zeromus** [comment]
Uhhhh I'll take a stab at solving the scond_wait_timeout within the next couple of days. Even a broken implementation will be better than nothing. It would be nice if someone could confirm whether retroarch is actually depending on that function right now (if it is, then we can't solve it like alcaro suggested)

**Comment by Alcaro** [comment]
```
alcaro@stacked ~/Desktop/minir/cores/RetroArch $ grep -r scond_wait_timeout
frontend/drivers/platform_gx.c:      scond_wait_timeout(gx_device_cond, gx_device_cond_mutex, 1000000);
audio/drivers/coreaudio.c:      if (write_avail == 0 && !scond_wait_timeout(
libretro-common/include/rthreads/rthreads.h: * scond_wait_timeout:
libretro-common/include/rthreads/rthreads.h:bool scond_wait_timeout(scond_t *cond, slock_t *lock, int64_t timeout_us);
libretro-common/rthreads/rthreads.c: * scond_wait_timeout:
libretro-common/rthreads/rthreads.c:bool scond_wait_timeout(scond_t *cond, slock_t *lock, int64_t timeout_us)
Binary file obj-unix/libretro-common/rthreads/rthreads.o matches
Binary file obj-unix/tasks/task_save.o matches
Binary file obj-unix/gfx/video_thread_wrapper.o matches
tasks/task_save.c:         scond_wait_timeout(save->cond, save->cond_lock,
Binary file retroarch matches
gfx/video_thread_wrapper.c:         if (!scond_wait_timeout(thr->cond_cmd, thr->lock, delta))
```

platform_gx is Gamecube/Wii, and CoreAudio looks like some OSX stuff, so we can ignore that. The rthreads files obviously don't depend on themselves, and the object files are irrelevant as well.

That leaves task_save and video_thread_wrapper. I don't know which of those are unused on Windows, which can safely remove the timeout, which can safely and easily be replaced with a semaphore, and which are dealbreakers.

And the cores, of course. I don't know which of them use this function. Or threads at all.

**Comment by zeromus** [comment]
OK, this passed my cursory tests. I'm pretty good at this kind of stuff, but I'm not confident I did it correctly. Nonetheless I think it's still better than what we had before.  Unless I broke the core algorithm. As planned, I extended the core algorithm to incorporate a sense of timeouts. The timeout is just infinite for scond_wait().

Notably I employ timeGetTime now (how else am I going to simulate a timeout?)
This means winmm.lib linking, which--well, this is meant for realtime game software, so there's no reason to care.

Regarding comment style, I am blind to comment styles, so I have no rhyme or reason for what I do.

**Comment by Alcaro** [comment]
> (how else am I going to simulate a timeout?)

QueryPerformanceCounter? GetSystemTimeAsFileTime? GetTickCount? Personally I'd go for QPC, the others (including tGT) seem to have 16ms granularity according to a quick googling. ...but on the other hand, maybe the WFSO timeouts are also 16ms, higher granularity than WFSO is just a waste of time.

> Regarding comment style, I am blind to comment styles, so I have no rhyme or reason for what I do.

We can clean it up later, no big deal.

> \+ /* Someone asking for a 0 timeout clearly wants no timeout. */

I'm pretty sure a standards compliant cv can't be signalled on a zero timeout, other than spuriously.

Everything looks fine to me, except this thing:

> \+   printf("%d\n",cond->wakens);

**Comment by zeromus** [comment]
I dunno which of those functions are on win9x. 16ms isn't useful for our purposes, as it's on the order of a frame time which will break any core which relied on this for some reason. I don't think putting a 16ms granularity timeout in this library is going to be doing anyone any favors.

So, I agree that a 0 timeout on a pthread CV is basically nonsense. Just to be clear, I assume someone calling scond_wait_timeout with a 0 timeout is probably meaning that they want to yield if the CV couldnt be obtained. At least that's what I want as a windows programmer. If I asked PThreads, he would say "OK, then explicitly yield". Fine, but rthreads doesnt have a yield API -- so this functionality wouldn't be possible right now

That's another problem to solve another day.

However in general, I loathe pthreads, and would prefer to not have to tell users of rthreads that they have to do bend-over-backwards pthreads shenanigans when we have an opportunity to refine the interface (and have, in this instance, actually already done that by following a more normal timeout interval based API instead of absolute time). But that isn't my call.

You'll see in my next commit that I decided to treat a 0 timeout as nonsense in windows too and say it just always times out. I feel good about that.

**Comment by bparker06** [comment]
QPC should work on 9x but give these a read also:

http://www.virtualdub.org/blog/pivot/entry.php?id=106

https://blogs.msdn.microsoft.com/oldnewthing/20050902-00/?p=34333

**Comment by Alcaro** [comment]
> I don't think putting a 16ms granularity timeout in this library is going to be doing anyone any favors.

If the WaitForSingleObject timeout runs on a 16ms granularity, higher granularity for our timer query function won't help.

The entire idea of relying on timeouts creeps me out.

> if the CV couldnt be obtained

That's not how cvs work. If the wakeup isn't spurious or timeout, it's because a signal was sent while the thread was sleeping; if the sleep duration is zero, it's kinda hard to hit it.

I'm not sure if we need the return value at all, proper cv use requires rechecking condition and timer no matter what the return is. But the existing API has a return value, so let's just keep it.

> If I asked PThreads, he would say "OK, then explicitly yield".

I believe a standards compliant implementation is allowed to instantly return failure and pretend it instantly re-locked the mutex.

In practice, it probably does yield the mutex, sounds easier to implement.

> rthreads doesnt have a yield API

Not sure if we need one. But if we do, it should be called sthread_yield() or something; don't overload it on an existing function, that's just silly.

> However in general, I loathe pthreads

Death to cv, I cannot think of any plausible situation where cv would work better than a semaphore. Maybe cv made sense when implemented directly in the OS scheduler, but that was long ago.

pthread has a lot of features, but many (if not most) of them seem stupid. Why do both normal and errorchecking mutexes exist, and why can I only have one of robust, errorcheck and recursive?

> following a more normal timeout interval based API instead of absolute time

I'd prefer not doing that. I'd rather see this one be a portable implementation of the useful parts of pthreads, modulo function names.

But that's not relevant here.

 

tldr: I can see a couple of possible improvements to the sthread API, but they belong in separate issues or PRs. If you're ready to merge this, then so am I.

**Comment by zeromus** [comment]
I have no idea what granularity WaitForSingleObject runs at. If we have evidence it runs at crappy granularity, then we can use a slightly less obnoxious timing mechanism.

You're right, thats not how pthread (which I don't like) CV work. I said that's how someone calling scond_wait_timeout() would expect it to work. My evidence for this is that that's how _I_ expected it to work. 

Regarding whether you follow a "portable implementation of the useful parts of pthreads" or use a customized API--I've got to be blunt about this. You're wrong. It's an extraordinary waste of resources to try porting an API which was designed around how someone happened to program their unix scheduler to every platform under the sun. You should be designing something portable, not spending 100s of man hours struggling to only very crudely nail a portable subset of something not designed to be portable.  

Not that I'm advising this, as it suffers from some of the same problems as pthreads (though not nearly as many) but for comparison an alternative would be looking at what newer c++ or C standards implement--those have actually been thought about in terms of suitability for porting, as well as to some extent a 'clean slate' approach

To that end, I have no problem with making scond_wait_timeout work however I think makes sense--and in this case that includes interpreting a 0 timeout in the only sensible way: as different from pthread's (which I don't care about) cvars.

I'm ready to merge this also. If it causes problems in the wild, I'll be on call to fix them.

**Comment by Alcaro** [comment]
Then let's do it, before Twinaphex starts questioning us again.

**Comment by Alcaro** [comment]
> spending 100s of man hours struggling to only very crudely nail a portable subset of something not designed to be portable

[hmm](https://github.com/libretro/RetroArch/pull/4392)

> an alternative would be looking at what newer c++ or C standards implement

C++11 condition variables support both relative and absolute timeouts.

C11 implements only the absolute one. POSIX is absolute.

java.lang.Object.wait is relative only. C# [is relative](https://msdn.microsoft.com/en-us/library/syehfawa.aspx). Rust [is relative](https://doc.rust-lang.org/std/sync/struct.Condvar.html). [Windows native](https://msdn.microsoft.com/en-us/library/windows/desktop/ms686301(v=vs.85).aspx) is relative. [Qt is relative.](http://doc.qt.io/qt-4.8/qwaitcondition.html)

You win this round, relative is more common.

> This means winmm.lib linking, which--well, this is meant for realtime game software, so there's no reason to care.

For RetroArch, that's fine.

For cores, it's less obvious. And I'm pretty sure libretro-common is supposed to be usable for non-libretro purposes, even if it's obviously not its primary purpose.

Not sure how much that matters. Probably not much.

**Comment by bparker06** [comment]
@zeromus Some Windows-specific code is now failing to build in the "new" vc2005 debug target (but not release):

http://p.0bl.net/13788

Any ideas?

**Comment by zeromus** [comment]
winmm.lib is only specified for `Release|Win32` configuration. Need to do properties on the project, change 'configuration' to "all configurations", and reset the linker additional dependencies to `msimg32.lib winmm.lib`. I dont have vs2005 installed so I can't do it safely without upgrading the stuff

**Comment by bparker06** [comment]
Thanks, I just edited the file manually to add it where it was missing.
